source/custom/nimxlcterminaladaptor_restricted_converters.h

```diff
@@ -16,35 +17,81 @@ namespace nimxlcterminaladaptor_restricted_grpc {
 // Add underscore to usings so they don't conflict with including files in the same namespace.
 namespace pb_ = ::google::protobuf;
 
-bool nierr_noOpReallocJson(struct nierr_Status * s, uint32_t size);
-
-void nierr_Status_initialize(struct nierr_Status * status);
-
+template <typename TServerContext>
 struct NIErrStatusOutputConverter {
-  NIErrStatusOutputConverter()
+  NIErrStatusOutputConverter(TServerContext *serverContext)
   {
+    context = serverContext;
     nierr_Status_initialize(&status);
+    url_table_initialize();
+  }
+
+  ~NIErrStatusOutputConverter()
+  {
+    nierr_Status_jsonFree(&status);
   }
 
   nierr_Status* operator&()
   {
     return &status;
   }
 
+  std::string url_encode(const char* input) const
+  {
+    std::string encoded(input);
+    // Replace the following characters to their equivalent '%xx' hex code
+    for (size_t position = 0; position < encoded.length(); ) {
+      if (urlLookupTable.test(encoded[position])) {
+        position++;
+      }
+      else {
+        char character[4];
+        ::_snprintf(character, 4, "%%%2x", static_cast<int>(encoded[position]));
```

Comment:
`_snprintf` is MSVC-specific:

```
[ 79%] Building CXX object CMakeFiles/UnitTestsRunner.dir/source/tests/unit/nimxlc_terminal_adaptor_converters_tests.cpp.o
In file included from /home/runner/work/grpc-device/grpc-device/source/tests/unit/nimxlc_terminal_adaptor_converters_tests.cpp:3:
/home/runner/work/grpc-device/grpc-device/./source/custom/nimxlcterminaladaptor_restricted_converters.h: In member function ‘std::string nimxlcterminaladaptor_restricted_grpc::NIErrStatusOutputConverter<TServerContext>::url_encode(const char*) const’:
/home/runner/work/grpc-device/grpc-device/./source/custom/nimxlcterminaladaptor_restricted_converters.h:49:11: error: ‘::_snprintf’ has not been declared; did you mean ‘vsnprintf’?
   49 |         ::_snprintf(character, 4, "%%%2x", static_cast<int>(encoded[position]));
      |           ^~~~~~~~~
      |           vsnprintf
```

Try changing this to use `snprintf`. [Microsoft's docs](https://learn.microsoft.com/en-us/cpp/c-runtime-library/reference/snprintf-snprintf-snprintf-l-snwprintf-snwprintf-l?view=msvc-170) say it's supported in C++:

> Beginning with the UCRT in Visual Studio 2015 and Windows 10, snprintf is no longer identical to _snprintf. The snprintf function behavior is now C99 standard conformant.

Reply:
Fixed.